src/main/java/org/apache/flink/benchmark/BufferDebloatedCheckpointTimeBenchmark.java

```diff
@@ -0,0 +1,261 @@
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.net.URI;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.apache.flink.api.common.eventtime.WatermarkStrategy.noWatermarks;
+
+/**
+ * The test verifies that the debloating kicks in and properly downsizes buffers. In the end the checkpoint should
+ * take ~2(number of rebalance) * DEBLOATING_TARGET.
+ *
+ * <p>Some info about the chosen numbers:
+ * <ul>
+ * <li>The minimal memory segment size is decreased (256b) so that the scaling possibility is higher. Memory segments
+ *    start with 4kb</li>
+ * <li>A memory segment of the minimal size fits ~3 records (of size 64b), each record takes ~1ms to be processed by
+ *   the sink</li>
+ * <li>We have 2 (exclusive buffers) * 4 (parallelism) + 8 floating = 64 buffers per gate, with 300 ms debloating
+ *   target and ~1ms/record processing speed, we can buffer 300/64 = ~4.5 records in a buffer after debloating which
+ *   means the size of a buffer is slightly above the minimal memory segment size.</li>
+ *  <li>The buffer debloating target of 300ms means a checkpoint should take ~2(number of exchanges)*300ms=~600ms</li>
+ * </ul>
+ */
+@State(Scope.Thread)
+@OutputTimeUnit(MILLISECONDS)
+@BenchmarkMode(Mode.SingleShotTime)
+@Fork(
+		value = 3,
+		jvmArgsAppend = {
+				"-Djava.rmi.server.hostname=127.0.0.1",
+				"-Dcom.sun.management.jmxremote.authenticate=false",
+				"-Dcom.sun.management.jmxremote.ssl=false",
+				"-Dcom.sun.management.jmxremote.ssl"
+		})
+@Warmup(iterations = 4)
+@Measurement(iterations = 10)
+public class BufferDebloatedCheckpointTimeBenchmark {
+	public static final int JOB_PARALLELISM = 4;
+	public static final MemorySize START_MEMORY_SEGMENT_SIZE = MemorySize.parse("4 kb");
+	public static final MemorySize MIN_MEMORY_SEGMENT_SIZE = MemorySize.parse("256 b");
+	public static final Duration DEBLOATING_TARGET = Duration.of(300, ChronoUnit.MILLIS);
+	public static final int RECORD_SIZE = (int) MemorySize.parse("64b").getBytes();
+
+	public static void main(String[] args) throws RunnerException {
+		Options options =
+				new OptionsBuilder()
+						.verbosity(VerboseMode.NORMAL)
+						.include(BufferDebloatedCheckpointTimeBenchmark.class.getCanonicalName())
+						.build();
+
+		new Runner(options).run();
+	}
+
+	@Benchmark
+	public void unalignedCheckpoint(DebloatedCheckpointEnvironmentContext context)
+			throws Exception {
+		final CompletableFuture<String> checkpoint = context.miniCluster.triggerCheckpoint(context.jobID);
+		checkpoint.get();
+	}
+
+	@State(Scope.Thread)
+	public static class DebloatedCheckpointEnvironmentContext {
+		public JobID jobID;
+		public StreamExecutionEnvironment env;
+		public MiniCluster miniCluster;
+
+		@Setup(Level.Invocation)
```

Comment:
I used `Level.Invocation` so that the buffer debloating happens before each benchmark method. It's the matter of what we test in each benchmark.

Initially I thought we wanted to have tens of buffer debloating processes. However if you think we should test tens of checkpoints after a single buffer debloating I am good with that as well. It actually makes sense.

Reply:
The purpose of the benchmark is to check effectiveness of the debloating, meassured as aligned checkpointing time. Multiple iterations/forks/invocations are only a mean to provide as stable results as possible. As long as results are trustworthy and stable, I don't mind if the debloating/setup happens once per fork (`Trial`), or once per invocation. I think the former one will provide more stable results, as we will even have a couple of "warm up" checkpoints before the measurement starts.